File: Binary Search/69. Sqrt(x)/Sqrt(x).cpp
#include <iostream>
using namespace std;

class Solution {
public:
    // Binary search
    // Time complexity: O(logn)
    // Space complexity: O(1)
    int mySqrt(int x) {
        if (x < 2) return x;
        int start = 1, end = x, ans = 0;
        while (start <= end) {
            int mid = start + (end - start) / 2;
            if (mid <= x / mid) {
                ans = mid;
                start = mid + 1;
            } else {
                end = mid - 1;
            }
        }
        return ans;
    }
};

int main() {
    Solution solution;
    int x=1000;
    cout << "The square root of " << x << " is " << solution.mySqrt(x) << endl;
    return 0;
}

File: Binary Search/69. Sqrt(x)/Sqrt(x).py
class Solution:
    #Approach: Binary Search
    #Time Complexity: O(logn)
    #Space Complexity: O(1)
    def mySqrt(self, x: int) -> int:
        start=0
        end=x
        mid=int(start+(end-start)/2)
        ans=0
        while(start<=end):
            if(mid*mid==x):
                return mid
            elif(mid*mid<x):
                ans=mid
                start=mid+1
            else:
                end=mid-1
            mid=int(start+(end-start)/2)
        return ans

if __name__ == "__main__":
    sol = Solution()
    x = 10000
    print(f"The square root of {x} is {sol.mySqrt(x)}")
